Source/WebCore/CMakeLists.txt

```diff
@@ -2608,6 +2608,12 @@ if (PORT STREQUAL GTK OR PORT STREQUAL WPE)
     )
 endif ()
 
+if (ARM_THUMB2_DETECTED)
+    WEBKIT_ADD_TARGET_CXX_FLAGS(WebCore
+        -Wno-cast-align
```

Comment:
Documentation of this warning:

```
       -Wcast-align
           Warn whenever a pointer is cast such that the required alignment of the target is increased.  For example, warn if a "char *" is cast to an "int
           *" on machines where integers can only be accessed at two- or four-byte boundaries.
```

Are you sure it's safe to turn this off? Looks like it indicates serious bugs? We could alternatively use `-Wcast-align=strict` to surface the warnings on other architectures so they're not limited to just 32-bit ARM.

Reply:
The problem is that GObject is 4-byte aligned, but the second you put anything interesting inside a struct on armv7, it becomes 8-byte aligned.

I would love to have a more robust solution to this, but we can't exactly make GObject* declare its alignment to be bigger. Unfortunately I also don't know of a way to force a greater alignment for imported types.

I suppose we could define a helper to do these downcasts for us, and have it squash this warning (or ASSERT the correct alignment), would that be better? It would probably churn a lot of code.

cast-align=strict seems like a good option if we make that change.

Reply:
I don't know what we should do about this.

Reply:
I think the real question is not about the _warning_, but about why the alignment mistake is not a real problem. If it is a real problem, then we can start using a separate memory block for the things that need more alignment instead of putting the data members directly into the class derived from `GObject`. If it’s not a problem it would be good to understand better why, and then I _do_ think the helper functions would be a good strategy to avoid turning off the warning for all the nearby code that might benefit from it just to quiet the.

Reply:
What is the failure mode? So if we ever allocated a GObject at 4 byte aligned address, and then cast it to a subclass that needs to be 8 byte aligned, and we'd access a member, we'd likely receive a SIGBUS and be killed?

> but we can't exactly make GObject* declare its alignment to be bigger

Why not? That seems to me like the elegant solution.

Reply:
You can't change GObject ABI ever. That's one level short of changing glibc ABI; think great social upheaval, riots in the streets, possibly the end of times. Got to just make the best of the way things are.

> What is the failure mode? So if we ever allocated a GObject at 4 byte aligned address, and then cast it to a subclass that needs to be 8 byte aligned, and we'd access a member, we'd likely receive a SIGBUS and be killed?

I'm not sure; I've never actually understood alignment requirements. But GObjects get cast between types very frequently, so that sounds like the sort of thing that would happen constantly, and we're not seeing bug reports about SIGBUS on ARM. Then again, probably there are very few users testing WebKitGTK on 32-bit ARM.

Reply:
We don’t need to debate this here, clearly the WebKit project is not in change of glib, but for the record, this is absolutely not proposing “changing GObject ABI”, what we are proposing is correctly annotating it so the compiler correctly understands alignment that is the actual GObject ABI. The fact that they are always allocated aligned is part of the ABI! We’ve tackled so many of these types problems in the past on other platforms (both Apple ones and others); this is very familiar territory for the project. To make these warnings work properly in WebKit we should figure out the most expedient approach; it would be great to get the warnings for all the non-GObject things, so it would be good to find a way to deal with it without turning off the warning entirely.

Reply:
Pretty sure changing the alignment (e.g. by using `_Alignas`) would be an ABI break? It's not hard to make changes in GLib, so long as those changes don't break the ABI. But if the change would require rebuilding software that depends on GLib, the change will be rejected.

Reply:
Like I said, this is not where we’d make those decisions. I’m pretty sure such a change would not require rebuilding.